### PR TITLE
Image mapper picker

### PR DESCRIPTION
**Pull request by jourdain**

### CODE REVIEW

File: Sources/Rendering/Core/ImageMapper/index.js [hunk 1]
```diff
@@ -58,6 +59,79 @@ function vtkImageMapper(publicAPI, model) {
   };
 
   publicAPI.getIsOpaque = () => true;
+
+  publicAPI.intersectWithLineForPointPicking = (p1, p2, snap = false) => {
+    const imageData = publicAPI.getInputData();
+    const extent = imageData.getExtent();
+    const normal = [0, 0, 0];               // Not as easy as that => Ken help me...
+    normal[model.currentSlicingMode] = 1;   // Not as easy as that => Ken help me...
```
Comment (jourdain) [inline Sources/Rendering/Core/ImageMapper/index.js]: @martinken There is a method on the image data to get the matrix for the axis. I think we should be able to get the one we are looking for from it, but I'm more than rusty on that subject.

Can you fix/improve those 2 functions?

**Comment by jourdain** [comment]
@martinken I let you merge it if you feel that's good. Otherwise let me know how to improve the code for the matrix.

Thanks

File: Sources/Common/Core/MatrixBuilder/index.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+    vec3.set(src, originDirection[0], originDirection[1], originDirection[2]);
+    vec3.set(dst, targetDirection[0], targetDirection[1], targetDirection[2]);
+    vec3.normalize(src, src);
+    vec3.normalize(dst, dst);
+    const cosAlpha = vec3.dot(src, dst);
+    if (cosAlpha > 1) {
+      return this;
+    }
+
```
Comment (martinken) [inline Sources/Common/Core/MatrixBuilder/index.js]: Should be >= 1 here

File: Sources/Filters/Sources/ConeSource/index.js [hunk 3]
```diff
@@ -60,6 +61,13 @@ function vtkConeSource(publicAPI, model) {
...
 
+    // Apply any tranformation to the points coordinates
+    vtkMatrixBuilder
+      .buildFromRadian()
+      .translate(...model.center)
+      .rotateFromDirections([1, 0, 0], model.direction)
+      .apply(points);
+
     // FIXME apply tranform
     dataset = vtkPolyData.newInstance();
```
Comment (martinken) [inline Sources/Filters/Sources/ConeSource/index.js]: Remove fixme comment?

File: Sources/Rendering/Core/ImageMapper/index.js (hunk 1 above)
Comment (martinken) [reply Sources/Rendering/Core/ImageMapper/index.js]: // I am assuming that p1, p2 have been converted into model coordinates
// via the prop's matrix

publicAPI.intersectWithLineForPointPicking = (p1, p2, snap = false) => {
     const imageData = publicAPI.getInputData();
     const extent = imageData.getExtent();

     const ijknormal = [0, 0, 0];               // Not as easy as that => Ken help me...
     ijknormal[model.currentSlicingMode] = 1;   // Not as easy as that => Ken help me...

     const sliceNum = 
      ijknormal[0]*model.xSlice +
      ijknormal[1]*model.ySlice +
      ijknormal[2]*model.zSlice;

     ijkorigin[0] = ijknormal[0] ? sliceNum : extent[0];
     ijkorigin[1] = ijknormal[1] ? sliceNum : extent[2];
     ijkorigin[2] = ijknormal[2] ? sliceNum : extent[4];

     imageData.indexToWorld(ijkorigin, wcorigin);

     ijkorigin[0] = ijknormal[0] ? sliceNum + 1 : extent[0];
     ijkorigin[1] = ijknormal[1] ? sliceNum + 1 : extent[2];
     ijkorigin[2] = ijknormal[2] ? sliceNum + 1 : extent[4];

     imageData.indexToWorld(ijkorigin, wcpt2);

     wcnormal = [
      wcpt2[0] - mcorigin[0],
      wcpt2[1] - mcorigin[1],
      wcpt2[2] - mcorigin[2]];
    vec3.normalize(wcnormal);

Comment (jourdain) [reply Sources/Rendering/Core/ImageMapper/index.js]: Great! I like the way you do it. I'll do the edits and let you know for a final review.

Thanks again!

File: Sources/Rendering/Core/ImageMapper/index.js [hunk 4]
```diff
@@ -58,6 +61,98 @@ function vtkImageMapper(publicAPI, model) {
...
+    }
+    return null;
+  };
+
+  publicAPI.intersectWithLineForCellPicking = (p1, p2, snap = false) => {
+    const imageData = publicAPI.getInputData();
+    const extent = imageData.getExtent();
+
+    // Slice origin
+    const ijk = [model.xSlice + 0.5, model.ySlice + 0.5, model.zSlice + 0.5];
```
Comment (jourdain) [inline Sources/Rendering/Core/ImageMapper/index.js]: @martinken do you agree for the 0.5 shift to deal with cell ijk?

File: Sources/Rendering/Core/ImageMapper/index.js [hunk 5]
```diff
@@ -58,6 +61,98 @@ function vtkImageMapper(publicAPI, model) {
...
+        || ijk[2] < extent[4] || ijk[2] > extent[5]) {
+        return null;
+      }
+
+      // Get closer ijk
+      if (snap) {
+        ijk[0] = Math.round(ijk[0] - 0.5);
+        ijk[1] = Math.round(ijk[1] - 0.5);
+        ijk[2] = Math.round(ijk[2] - 0.5);
+      }
```
Comment (jourdain) [inline Sources/Rendering/Core/ImageMapper/index.js]: Same here @martinken?

**Comment by jourdain** [comment]
@martinken I think I'm ready to merge if you agree with my 0.5 logic for the cell IJK

**Pull request merged**